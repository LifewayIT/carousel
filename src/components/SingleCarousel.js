import React, { useRef, useLayoutEffect } from 'react';
import styled from 'styled-components';
import {
  alignAtCenter,
  getFirstFocusableElement,
  scrollTo
} from './utils';
import { useResizeEffect } from '../hooks/layout';

const Container = styled.div`
  overflow: hidden;

  display: flex;
  flex-direction: row;
  align-items: stretch;

  & > * {
    flex: 0 0 auto;
  }
`;

const getTiles = container => Array.from(container.children);
const getTile = (container, num) => getTiles(container)[num]

const scrollIntoView = (container, el, smooth = true) => {
  if (!container || !el) return;

  scrollTo(container, alignAtCenter(container, el), smooth);
}



const SingleCarousel = ({ selected, onSelect, children }) => {
  const containerRef = useRef();
  
  useLayoutEffect(() => {
    if (!containerRef.current) return;
    const container = containerRef.current;

    const selectedTile = getTile(container, selected);

    scrollIntoView(container, selectedTile);
  }, [selected]);

  useResizeEffect(containerRef, () => {
    scrollIntoView(containerRef.current, getTile(containerRef.current, selected), false);
  });

  const scrollFocusedElementIntoView = (evt) => {
    if (!containerRef.current) return;
    const container = containerRef.current;

    const tiles = getTiles(container);
    const focusedNum = tiles.findIndex(tile => tile.contains(evt.target));

    scrollIntoView(container, tiles[focusedNum]);
    onSelect(focusedNum);
  };

  const onKeyDown = (e) => {
    const container = containerRef.current;
    if (e.key === 'ArrowLeft') {
      const prev = Math.max(selected - 1, 0);
      onSelect(prev);
      getFirstFocusableElement(getTile(container, prev))?.focus();
    } else if (e.key === 'ArrowRight') {
      const next = Math.min(selected + 1, React.Children.count(children) - 1);
      onSelect(next);
      getFirstFocusableElement(getTile(container, next))?.focus();
    }
  };

  return (
    <Container ref={containerRef} onFocus={scrollFocusedElementIntoView} onKeyDown={onKeyDown}>
      {children}
    </Container>
  );
};

export default SingleCarousel;
