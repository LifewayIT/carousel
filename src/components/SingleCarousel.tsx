import React, {
  FocusEventHandler,
  KeyboardEventHandler,
  ReactElement,
  ReactNode,
  RefObject,
  useLayoutEffect,
  useRef
} from 'react';
import styled from 'styled-components';
import { useResizeEffect } from '../hooks/layout';
import {
  alignAtCenter,
  getFirstFocusableElement,
  scrollTo
} from './utils';

const Container = styled.div`
  overflow: hidden;

  display: flex;
  flex-direction: row;
  align-items: stretch;

  & > * {
    flex: 0 0 auto;
  }
`;

const getTiles = (container: HTMLElement) => Array.from(container.children) as HTMLElement[];
const getTile = (container: HTMLElement, num: number): HTMLElement | undefined => getTiles(container)[num];

const scrollIntoView = (container: HTMLElement | undefined, el: HTMLElement | undefined, smooth = true) => {
  if (!container || !el) return;

  scrollTo(container, alignAtCenter(container, el), smooth);
};


type HookProps = {
  selected: number;
  onSelect: (nextSelected: number) => void;
  children?: ReactNode;
};

const useKeepSelectedTileInView = (containerRef: RefObject<HTMLElement>, selected: number) => {
  useLayoutEffect(() => {
    if (!containerRef.current) return;
    const container = containerRef.current;

    const selectedTile = getTile(container, selected);

    scrollIntoView(container, selectedTile);
  }, [selected]);

  useResizeEffect(containerRef, () => {
    if (!containerRef.current) return;
    scrollIntoView(containerRef.current, getTile(containerRef.current, selected), false);
  });
};

const useSelectOnFocus = (containerRef: RefObject<HTMLElement>, onSelect: (selected: number) => void) => {
  const scrollFocusedElementIntoView: FocusEventHandler = (evt) => {
    if (!containerRef.current) return;
    const container = containerRef.current;

    const tiles = getTiles(container);
    const focusedNum = tiles.findIndex(tile => tile.contains(evt.target));

    scrollIntoView(container, tiles[focusedNum]);
    onSelect(focusedNum);
  };

  return {
    onFocus: scrollFocusedElementIntoView
  };
};

const useSelectWithArrowKeys = (containerRef: RefObject<HTMLElement>, { selected, onSelect, children }: HookProps) => {
  const onKeyDown: KeyboardEventHandler = (e) => {
    const container = containerRef.current;
    if (e.key === 'ArrowLeft') {
      const prev = Math.max(selected - 1, 0);
      onSelect(prev);

      if (container) {
        getFirstFocusableElement(getTile(container, prev))?.focus();
      }
    } else if (e.key === 'ArrowRight') {
      const next = Math.min(selected + 1, React.Children.count(children) - 1);
      onSelect(next);

      if (container) {
        getFirstFocusableElement(getTile(container, next))?.focus();
      }
    }
  };

  return { onKeyDown };
};

type Props = {
  /** the index of the child that is currently selected. defaults to 0 */
  selected?: number;
  /** handler for when a child is selected */
  onSelect?: (nextSelected: number) => void;
  /** the children to render */
  children?: ReactNode;
};

const SingleCarousel = ({ selected = 0, onSelect = () => undefined, children }: Props): ReactElement => {
  const containerRef = useRef<HTMLDivElement>(null);

  useKeepSelectedTileInView(containerRef, selected);
  const focusProps = useSelectOnFocus(containerRef, onSelect);
  const keyProps = useSelectWithArrowKeys(containerRef, { selected, onSelect, children });

  return (
    <Container ref={containerRef} {...focusProps} {...keyProps}>
      {children}
    </Container>
  );
};

export default SingleCarousel;
