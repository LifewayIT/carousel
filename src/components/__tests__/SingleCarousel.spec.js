import React from 'react';
import { render } from '@testing-library/react';
import { SingleCarousel } from '../SingleCarousel';

test('matches snapshot', () => {
  const { container } = render(
    <SingleCarousel className="test-class" selected={0}>
      <div id="children" />
    </SingleCarousel>
  );

  expect(container.firstChild).toMatchSnapshot();
});
