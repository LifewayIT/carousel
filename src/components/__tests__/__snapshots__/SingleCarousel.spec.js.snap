// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="lwc-single-carousel test-class"
>
  {children}
</div>
`;
